README.md
# STRANDS User Interfaces

User interfaces for the robots, displayed using a web browser.

# Installation

You can add the STRANDS UI source to your catkin workspace (if it's not there already) using `wstool`:
```bash
wstool merge https://raw.github.com/hawesie/strands_ui/master/strands_ui_rosinstall.yaml
wstool update strands_ui twitter_bootstrap
```

`rosdep` should give you most requirements as usual (e.g. `rosdep install --from-paths src --ignore-src --rosdistro groovy -y -r`), but we don't yet have a rosdep for [web.py](http://webpy.org) which is required by `strands_webserver`. This can be installed using, e.g., 

```bash
sudo apt-get install python-webpy
```
or
```
sudo pip install web.py
```

You can then run `catkin_make` as usual.


# strands_webserver

The `strands_webserver` is a node which both acts as a webserver and a ROS node which can receive input from other nodes. This allows these other nodes to control what the webserver displays and receive feedback from user interaction with the pages.

## Running

To run the server, first run rosbrige if you don't have it running already.
```
roslaunch rosbridge_server rosbridge_websocket.launch
```
Then run the server
```bash
rosrun strands_webserver strands_webserver 
```

This will launch the webserver on localhost port 8090, so browse to [http://localhost:8090](http://localhost:8090) and you should see something like
```text
Ready; display = 1
```

## Displaying content

The webserver manages individual *displays* separately, one for each request to its URL. These can be published to individual by specifying their number, or you can just send 0 to publish to all of them.

Interaction with `strands_webserver` typically happens by publishing instructions to it for display. These can either be URLs or bits of html which will be injected into its main page. The URLs are loaded in a full size iframe. The bits of html are injected as the `innerHTML` of the `content` node in the main page (`strands_webserver/data/main.html`).

The example below shows how to publish a simple html page to the webserver, both a normal URL (given that it is happy to appear in an iframe) then a page defined locally.
```python
#! /usr/bin/env python

import sys
import os
import roslib
import rospy

import strands_webserver.client_utils

if __name__ == '__main__':
	rospy.init_node("strands_webserver_demo")
	# The display to publish on, defaulting to all displays
	display_no = rospy.get_param("~display", 0)	

	# display a start-up page
	strands_webserver.client_utils.display_url(display_no, 'http://strands-project.eu')

	# sleep for 2 seconds
	rospy.sleep(5.)

	# tell the webserver where it should look for web files to serve
	http_root = os.path.join(roslib.packages.get_pkg_dir("strands_webserver"), "data")
	strands_webserver.client_utils.set_http_root(http_root)

	# start with a basic page pretending things are going normally
	strands_webserver.client_utils.display_relative_page(display_no, 'example-page.html')
```

To display arbitrary HTML within the the `content` node in the main page (`strands_webserver/data/main.html`) you can use calls like the following
```python
strands_webserver.client_utils.display_content(display_no, '<p>Hello world</p>') 
```
Note that any `script` nodes will not be evaluated in this injected html. 

### Using the Webloader action server
The webloader action server offers 4 ways of manipultating webpages shown by the webserver (all of this of course only works if you exclusivly use the webloader server to display pages):
* Loading relative or contant generated pages
* Reloading the current page
* Going back to the previous page
* Showing a temporary page for a specific number of seconds

Usage still requires your main programm to set
```
	http_root = os.path.join(roslib.packages.get_pkg_dir("my_package"), "directory")
```

Currently all these pages are sent to all displays. Possible future extensions could therefore include a history for each display.

Every of the above mentioned actions has their own action file. Reloading and going back are just empty actions. Loading a page reuires to specify `goal.relative=BOOL` to toggle between relative and content generated pages and `goal.page=STRING` is either the name of the relative page or the content for the generated page. For the temporary page you also have to specify `goal.timeout=INT` after which the previous page will be shown again.

This action server was developed to manage the user interface for the NHM deployment where people could trigger to tweet an image. This was displayed on a temporary page and afterwards the display was reset to the previous page. This had the advantage of showing arbitrary pages without care about what should be displayed afterwards.

## Auto-generated button pages

The `strands_webserver` package also provides functionality for automatically generating pages with buttons which can call services. Such a call might look like this
```python
	notice = 'Help me, I am <em>stuck</em>'
	buttons = [('No', 'trigger_pleading'), ('Sure', 'party_time')]
	service_prefix = '/caller_services'
	content = strands_webserver.page_utils.generate_alert_button_page(notice, buttons, service_prefix)	
	strands_webserver.client_utils.display_content(display_no, content) 
```
In this `notice` is the used to generate a large banner notice and `buttons` maps between a button label (e.g. No) and the service call which will be triggered when the button is pressed, e.g. (`'/caller_services/trigger_pleading'`). This service is of type `std_srvs/Empty`. Instead of `generate_alert_button_page` you can use `generate_button_page` which accepts arbitrary html instead of a banner notice.

The full example is as follows (also available as `strands_webserver/scripts/strands_webserver_demo.py`).

```python
#! /usr/bin/env python

import sys
import os
import roslib
import rospy

import strands_webserver.page_utils
import strands_webserver.client_utils
import std_srvs.srv 


def trigger_pleading(req):
      print 'please, please help'


def party_time(req):
      print 'woo, off I go baby'


if __name__ == '__main__':
	rospy.init_node("strands_webserver_demo")
	# The display to publish on, defaulting to all displays
	display_no = rospy.get_param("~display", 0)	

	# display a start-up page
	strands_webserver.client_utils.display_url(display_no, 'http://strands-project.eu')

	# sleep for 5 seconds
	rospy.sleep(5.)

	# tell the webserver where it should look for web files to serve
	http_root = os.path.join(roslib.packages.get_pkg_dir("strands_webserver"), "data")
	strands_webserver.client_utils.set_http_root(http_root)

	# start with a basic page pretending things are going normally
	strands_webserver.client_utils.display_relative_page(display_no, 'example-page.html')

	# sleep for 5 seconds
	rospy.sleep(5.)

	# now ask for help
	name = 'Help me, I am <em>stuck</em>'
	buttons = [('No', 'trigger_pleading'), ('Sure', 'party_time')]
	service_prefix = '/caller_services'
	content = strands_webserver.page_utils.generate_alert_button_page(name, buttons, service_prefix)	
	strands_webserver.client_utils.display_content(display_no, content) 
	rospy.Service('/caller_services/trigger_pleading', std_srvs.srv.Empty, trigger_pleading) 
	rospy.Service('/caller_services/party_time', std_srvs.srv.Empty, party_time) 
	rospy.spin()
```

## Includes

Need to describe where the webserver fetches javascript and css from, and the standard includes available.

# marathon_touch_gui

This package uses the `strands_webserver` to create an interface for the patroller during the marathon event. It's not very pretty, but it's a start. There is a main page (map, pause button) which is displayed using `strands_webserver.client_utils.display_relative_page` and two pages for recovery methods which are generated using `strands_webserver.page_utils.generate_alert_button_page`. These are wrapped up in `marathon_touch_gui.client` for ease of use. They can be called as follows:
```python
	# Setup -- must be done before other marathon_touch_gui calls
	marathon_touch_gui.client.init_marathon_gui()

	# Show the main page of the GUI
	marathon_touch_gui.client.display_main_page(displayNo)

	rospy.sleep(2)
	
	# All callback services should be under this prefix
	service_prefix = '/patroller'

	# Do something like this on bumper collision

	# when the human gives the 'ok' then /patroller/bumper_recovered is called
	# note that this service must be provided by some other node
	on_completion = 'bumper_recovered'
	marathon_touch_gui.client.bumper_stuck(displayNo, service_prefix, on_completion)

	rospy.sleep(2)

	# Do something like this on move_base failure

	# when the human gives the 'ok' then /patroller/robot_moved is called
	# note that this service must be provided by some other node
	on_completion = 'robot_moved'
	marathon_touch_gui.client.nav_fail(displayNo, service_prefix, on_completion)

	rospy.sleep(2)

	# Return to main page
	marathon_touch_gui.client.display_main_page(displayNo)
```
The full example is in `marathon_touch_gui/scripts/demo.py`

## Running

To run the marathon GUI, first launch `strands_webserver` plus rosbridge and the necessary additional publishers (with HOST_IP set to an external ip for your machine if you want this to be externally accessible, else leave blank for 127.0.0.1):
```bash
HOST_IP=10.0.11.158 roslaunch marathon_touch_gui marathon_gui_dependencies.launch
```
This will launch the webserver on localhost port 8090, so browse to [http://localhost:8090](http://localhost:8090).

Then you can call the `marathon_touch_gui` functions. To test you can cycle through them with 
```bash
rosrun marathon_touch_gui demo.py 
```
## Modal Dialog

By publishing a `strands_webserver/ModalDlg` message on the `/strands_webserver/modal_dialog` topic a modal dialog can be triggered hovering above any window display. This is useful for system notifications. The `strands_webserver/ModalDlg` is defined as follows:

```
string title
string content
bool show
```

The modal dialog can be closed by the user using a little close item in the top right corner, it can remotely be hidden, by setting `show` to false. Here is an example to display the dialog box:

```
rostopic pub /strands_webserver/modal_dialog strands_webserver/ModalDlg "title: 'Nice Title'
content: '<b>test</b> <a href="https://github.com/strands-project/aaf_deployment">AAF</a>'
show: true"
```

To hide it again, simply publish

```
rostopic pub /strands_webserver/modal_dialog strands_webserver/ModalDlg "title: ''
content: ''
show: false"
```

Both, title and content can contain valid HTML.




mary_tts/CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mary_tts
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

0.0.3 (2014-11-11)
------------------

0.0.4 (2014-11-11)
------------------
* 0.0.3
* updated changelogs
* Contributors: Jenkins

0.2.1 (2018-10-25)
------------------
* Merge pull request `#109 <https://github.com/strands-project/strands_ui/issues/109>`_ from francescodelduchetto/pull-req
  fix global variable declaration missing
* fix global variable declaration missing
* Contributors: Marc Hanheide, francescodelduchetto

0.2.0 (2018-10-16)
------------------
* Merge pull request `#108 <https://github.com/strands-project/strands_ui/issues/108>`_ from francescodelduchetto/pull-req
  Catch cancel_goal of the action and stop the player accordingly.
* goal preemption stops marytts speaking
* Contributors: Marc Hanheide, francescodelduchetto

0.1.1 (2018-08-13)
------------------
* Merge pull request `#107 <https://github.com/strands-project/strands_ui/issues/107>`_ from strands-project/sync_reply
  resolved sync issue with clearing queue
* resolved sync issue with clearing queue
* Contributors: Lindsey User, Marc Hanheide

0.1.0 (2017-06-30)
------------------

0.0.34 (2017-06-26)
-------------------

0.0.33 (2017-01-11)
-------------------

0.0.32 (2016-06-06)
-------------------

0.0.31 (2016-03-20)
-------------------

0.0.30 (2016-03-19)
-------------------

0.0.29 (2016-03-19)
-------------------

0.0.28 (2016-03-01)
-------------------

0.0.27 (2016-02-08)
-------------------

0.0.26 (2016-02-07)
-------------------
* added IP4 help
* Contributors: Marc Hanheide

0.0.25 (2015-05-10)
-------------------
* Allowing to set the locale via parameter using environment variable
* Contributors: Christian Dondrup

0.0.24 (2015-05-05)
-------------------

0.0.23 (2015-04-27)
-------------------

0.0.22 (2015-04-22)
-------------------

0.0.21 (2015-04-15)
-------------------

0.0.20 (2015-04-10)
-------------------
* Update marybridge.py
  Setting default for mary host to `localhost`
* Adding support to start mary on different machine
* Contributors: Christian Dondrup

0.0.19 (2015-03-31)
-------------------

0.0.18 (2015-03-28)
-------------------

0.0.17 (2015-03-24)
-------------------

0.0.15 (2015-03-09)
-------------------

0.0.11 (2014-11-22)
-------------------

0.0.10 (2014-11-20)
-------------------
* [mary_tts] Added troubleshooting info
* Contributors: Christian Dondrup

0.0.9 (2014-11-14)
------------------

0.0.8 (2014-11-13)
------------------
* Fixing a bug in the naming of the local en_US.
* Contributors: Christian Dondrup

0.0.7 (2014-11-13)
------------------
* Merge branch 'hydro-devel' of http://github.com/strands-project/strands_ui into hydro-devel
* Fixing `#52 <https://github.com/strands-project/strands_ui/issues/52>`_:
  * Enabling selection of `en_GB` locale
  * Setting `en_GB` as default locale because the default language is part of that
  Fixing `#50 <https://github.com/strands-project/strands_ui/issues/50>`_:
  * removed autostart from action server
  Additional bugfix:
  * If an empty string is pushied a warning will be diplayed and the ervice or action server report failure. This prevents mary from crashing when trying to say nothing.
* Adding correct permissions to marytts-5.0 install target
  Fixing `#53 <https://github.com/strands-project/strands_ui/issues/53>`_
* Contributors: Christian Dondrup

0.0.6 (2014-11-12)
------------------

0.0.5 (2014-11-11)
------------------
* 0.0.4
* updated changelogs
* 0.0.3
* updated changelogs
* Contributors: Jenkins

0.0.2 (2014-10-31)
------------------
* Updating dependencies.
  Preparing for release.
* moving human_help_manager service definition to human_help_manager pa…
  …ckage
* - made the Mary start script ROS compatible and installable
  - disabled file logging, closing https://github.com/strands-project/strands_hri/issues/6
* - made speak_webserver installable
  - made template and static dirs to be found
  - tidied up
* changed all ros_mary_tts to mary_tts
* changed package name
* Add 'mary_tts/' from commit 'ee851b0aa5851bc39fe6f13a6c3522d9f4783b74'
  git-subtree-dir: mary_tts
  git-subtree-mainline: 252851fb65fc7aa6c5439377d75a6899794efb36
  git-subtree-split: ee851b0aa5851bc39fe6f13a6c3522d9f4783b74
* Contributors: Bruno Lacerda, Christian Dondrup, Marc Hanheide


mary_tts/scripts/marybridge.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import httplib
import urllib

import StringIO

from Queue import *

import logging

from mary_tts.srv import *
import rospy
import roslib

import actionlib
import mary_tts.msg
import os

from pygame_managed_player.pygame_player import PyGamePlayer

import pygame.mixer as mixer


speakQueue = Queue()
replyQueue = Queue()
_finished_playing = False

class RosMary(object):
    def __init__(self, mary_client):
        self.mary_client = mary_client
        rospy.Service('ros_mary', ros_mary, self.speak)
        rospy.Service('ros_mary/set_voice', SetVoice, self.set_voice)
        rospy.Service('ros_mary/set_locale', SetLocale, self.set_locale)
        host = rospy.get_param("~host", "localhost")
        self.mary_client.set_host(host)

        # What voices are available?
        filelist = os.listdir(os.path.join(roslib.packages.get_pkg_dir("mary_tts"),
                                           "marytts-5.0/lib/"))
        print "Ready to speak."
        print "Locales:"
        self._locales = []
        for i in filelist:
            if "lang" in i:
                lang = i[13:i.find("-5.0")]
                if lang == 'en':  # English special case with two locales in one
                    lang = "en_GB"
                    print " - ", lang
                    self._locales.append(lang)
                    lang = "en_US"
                    print " - ", lang
                    self._locales.append(lang)
                else:
                    print " - ", lang
                    self._locales.append(lang)
        print "Voices: "
        self._voices = []
        for i in filelist:
            if "voice" in i and not "voices" in i:
                voice = i[6:i.find("-5.0")]
                print " - ", voice
                self._voices.append(voice)
        voice = rospy.get_param("~voice")
        locale = rospy.get_param("~locale")
        if locale not in self._locales:
            rospy.logwarn("Selected voice '%s' not available, using default!" % locale)
        else:
            self.mary_client.locale = locale
            print "Selected locale:", self.mary_client.locale
        if voice not in self._voices:
            rospy.logwarn("Selected voice '%s' not available, using default!" % voice)
        else:
            self.mary_client.voice = voice
            print "Selected voice:", self.mary_client.voice

    def speak(self, req):
        """ Speak service handler """
        if req.text == '':
            rospy.logwarn("mary_tts was asked to produce an empty string.")
            return False
        speakQueue.put(req.text)
        return True

    def set_voice(self, req):
        """ Voice setting service handle """
        if not req.voice_name in self._voices:
            rospy.logwarn("Trying to set voice to unknown:  %s", req.voice_name)
            return False
        self.mary_client.set_voice(req.voice_name)
        rospy.loginfo("Set voice to %s", self.mary_client.get_voice())
        return True

    def set_locale(self, req):
        """ Locale setting service handle """
        if not req.locale_name in self._locales:
            rospy.logwarn("Trying to set locale to unknown:  %s", req.locale_name)
            return False
        self.mary_client.set_locale(req.locale_name)
        rospy.loginfo("Set locale to %s", self.mary_client.get_locale())
        return True


#
# A simple MARY TTS client in Python, using pulseaudio for playback
#
# based on Code from Hugh Sasse (maryclient-http.py)
#
# 2013 by G. Bartsch. License: LGPLv3
class maryclient:

    def __init__(self):

        self.host = "127.0.0.1"
        self.port = 59125
        self.input_type = "TEXT"
        self.output_type = "AUDIO"
        self.audio = "WAVE_FILE"
        self.locale = "en_US"
        self.voice = "cmu-bdl-hsmm"
        self._action_name = "speak"
        self._as = actionlib.SimpleActionServer(self._action_name,
                                                mary_tts.msg.maryttsAction,
                                                execute_cb=self.execute_cb,
                                                auto_start=False)
        self._as.register_preempt_callback(self.preempt_cb)

        self._as.start()

    def set_host(self, a_host):
        """Set the host for the TTS server."""
        self.host = a_host

    def get_host(self):
        """Get the host for the TTS server."""
        return self.host

    def set_port(self, a_port):
        """Set the port for the TTS server."""
        self.port = a_port

    def get_port(self):
        """Get the port for the TTS server."""
        return self.port

    def set_input_type(self, t):
        """Set the type of input being
           supplied to the TTS server
           (such as 'TEXT')."""
        self.input_type = t

    def get_input_type(self):
        """Get the type of input being
           supplied to the TTS server
           (such as 'TEXT')."""
        return self.input_type

    def set_output_type(self, t):
        """Set the type of input being
           supplied to the TTS server
           (such as 'AUDIO')."""
        self.output_type = t

    def get_output_type(self):
        """Get the type of input being
           supplied to the TTS server
           (such as "AUDIO")."""
        return self.output_type

    def set_locale(self, a_locale):
        """Set the locale
           (such as "en_US")."""
        self.locale = a_locale

    def get_locale(self):
        """Get the locale
           (such as "en_US")."""
        return self.locale

    def set_audio(self, audio_type):
        """Set the audio type for playback
           (such as "WAVE_FILE")."""
        self.audio = audio_type

    def get_audio(self):
        """Get the audio type for playback
           (such as "WAVE_FILE")."""
        return self.audio

    def set_voice(self, a_voice):
        """Set the voice to speak with
           (such as "dfki-prudence-hsmm")."""
        self.voice = a_voice

    def get_voice(self):
        """Get the voice to speak with
           (such as "dfki-prudence-hsmm")."""
        return self.voice

    def generate(self, message):
        """Given a message in message,
           return a response in the appropriate
           format."""
        raw_params = {"INPUT_TEXT": message,
                "INPUT_TYPE": self.input_type,
                "OUTPUT_TYPE": self.output_type,
                "LOCALE": self.locale,
                "AUDIO": self.audio,
                "VOICE": self.voice,
                }
        params = urllib.urlencode(raw_params)
        headers = {}

        # Open connection to self.host, self.port.
        conn = httplib.HTTPConnection(self.host, self.port)

        # conn.set_debuglevel(5)

        conn.request("POST", "/process", params, headers)
        logging.warning('Watch out!' + params)

        response = conn.getresponse()
        if response.status != 200:
            print response.getheaders()
            raise RuntimeError("{0}: {1}".format(response.status,
                               response.reason))
        return response.read()

    def execute_cb(self, goal):
        # helper variables
        rospy.loginfo("action triggered: say %s", goal.text)
        if goal.text == '':
            rospy.logwarn("mary_tts was asked to produce an empty string.")
            self._as.set_aborted()
            return
        while not replyQueue.empty():
            replyQueue.get()
        speakQueue.put(goal.text)
        try:
            replyQueue.get(True, rospy.get_param("~timeout", 120))
        except Empty:
            rospy.logwarn("mary speach action failed; maybe took too long (more than 10 seconds), maybe pulse is broke.")
            self._as.set_succeeded(False)
            return
        rospy.loginfo("finished speaking...")
        self._as.set_succeeded()

    def preempt_cb(self):
        rospy.logwarn("Preemption request received, stopping mary speech")

    def is_preempt_requested(self):
        return self._as.is_preempt_requested()

def finished_playing_cb():
    global _finished_playing
    _finished_playing = True

if __name__ == "__main__":
    rospy.init_node('mary_tts')

    client = maryclient()

    client.set_locale("en_GB")
    client.set_voice("dfki-prudence-hsmm")

    rosmary = RosMary(client)

    player = PyGamePlayer(min_vol=1.0, max_vol=1.0, priority=1.0,
                          frequency=16000)
    player.stop_callback = finished_playing_cb

    while not rospy.is_shutdown():
        try:
            req = speakQueue.get(True, 1)
            rospy.loginfo("say " + req)
            the_sound = mixer.Sound(StringIO.StringIO(client.generate(req)))
            player.play_sound(the_sound, blocking=False)
            _finished_playing = False
            rate = rospy.Rate(5)
            while (not _finished_playing and not client.is_preempt_requested()):
                rate.sleep()
            player.stop()
            rospy.loginfo("played")
            replyQueue.put(True)
        except Empty:
            rospy.logdebug('nothing sent')
